### PR TITLE
Move storageclass config handling into cmd/config/storageclass

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Move storageclass config handling into cmd/config/storageclass

## Motivation and Context
Continuation of the changes done in PR #8351 to refactor,
add tests and move global handling into a more idiomatic
style for Go as packages.

## How to test this PR?
Please test the config migration and also the features in distributed, standalone setups related to storage class. 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [x] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#66bdffa3-8bb0-4680-93ba-ccfbedc44289) |
| mint-compression-dist-xl.sh | :x: [more...](#ee846248-3005-42ed-aad4-a2aeac3d31b1) |
<a name="ee846248-3005-42ed-aad4-a2aeac3d31b1"></a>
### 8360-6e0ea5c/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:31729
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 09ef3c262b31:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 45 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 2 minutes and 6 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 50 seconds
(8/14) Running minio-dotnet tests ... done in 39 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 12 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 51 seconds
(11/14) Running minio-js tests ... done in 53 seconds
(12/14) Running minio-py tests ... FAILED in 3 minutes and 41 seconds
{
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "args": {
    "object_name": "75de04e2-7473-4e95-8726-68c056208d9e",
    "response.status": 403,
    "bucket_name": "minio-py-test-9499020a-4408-434e-9d74-c012522cd7d6",
    "presigned_get_object_url": "http://72.28.97.53:31729/minio-py-test-9499020a-4408-434e-9d74-c012522cd7d6/75de04e2-7473-4e95-8726-68c056208d9e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20191005%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191005T192328Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=9f10b44e951b7e792cc112d4d3d81b0a79b5ff80b61f712958b2107bb8f2df1a",
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"440\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/xml\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-10-05T18-52-18Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15CAD6A33710B9F7\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Sat, 05 Oct 2019 19:23:36 GMT\"]}}",
    "response._body": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Request has expired</Message><Key>75de04e2-7473-4e95-8726-68c056208d9e</Key><BucketName>minio-py-test-9499020a-4408-434e-9d74-c012522cd7d6</BucketName><Resource>/minio-py-test-9499020a-4408-434e-9d74-c012522cd7d6/75de04e2-7473-4e95-8726-68c056208d9e</Resource><RequestId>15CAD6A33710B9F7</RequestId><HostId>90f304ca-6a99-44f2-a1f5-2af4fed1b49a</HostId></Error>",
    "response.reason": "Forbidden"
  },
  "status": "FAIL",
  "message": "AccessDenied: message: Access Denied",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1341, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2036, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1355, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "duration": 131
}

Executed 11 out of 14 tests successfully.
```
<a name="66bdffa3-8bb0-4680-93ba-ccfbedc44289"></a>
### 8360-6e0ea5c/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.54:31059
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 14aa53226171:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 3 seconds
(3/14) Running aws-sdk-php tests ... done in 51 seconds
(4/14) Running aws-sdk-ruby tests ... done in 16 seconds
(5/14) Running awscli tests ... done in 2 minutes and 24 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 17 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 19 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 49 seconds
{
  "args": {},
  "duration": 67886,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8360-6e0ea5c

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 8360-5154af4/gopath/src/github.com/minio
+ git -C 8360-5154af4/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 8360-5154af4/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 8360-5154af4/gopath/src/github.com/minio/minio fetch --quiet minio pull/8360/head:pr8360
+ git -C 8360-5154af4/gopath/src/github.com/minio/minio checkout --quiet pr8360
+ GO111MODULE=on
+ GOPATH=/mnt/nvme/mint-auto/8360-5154af4/gopath
+ make -C 8360-5154af4/gopath/src/github.com/minio/minio --quiet
go: creating work dir: mkdir /tmp/go-build670601214: no space left on device
go: creating work dir: mkdir /tmp/go-build290922554: no space left on device
go: creating work dir: mkdir /tmp/go-build965032891: no space left on device
go: creating work dir: mkdir /tmp/go-build267993528: no space left on device
go: creating work dir: mkdir /tmp/go-build032641898: no space left on device
/mnt/nvme/mint-auto/8360-5154af4/gopath/src/github.com/minio/minio/buildscripts/checkdeps.sh: line 76: cannot create temp file for here-document: No space left on device
/mnt/nvme/mint-auto/8360-5154af4/gopath/src/github.com/minio/minio/buildscripts/checkdeps.sh: line 77: cannot create temp file for here-document: No space left on device
/mnt/nvme/mint-auto/8360-5154af4/gopath/src/github.com/minio/minio/buildscripts/checkdeps.sh: line 76: cannot create temp file for here-document: No space left on device
/mnt/nvme/mint-auto/8360-5154af4/gopath/src/github.com/minio/minio/buildscripts/checkdeps.sh: line 77: cannot create temp file for here-document: No space left on device
go: creating work dir: mkdir /tmp/go-build777712313: no space left on device
make: *** [build] Error 1

```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#ab2dbdd7-5a56-4e5f-ac86-9cfc2b180346) |
| mint-large-bucket.sh | :x: [more...](#73af5f15-021a-4cde-8b8d-e532586d6ded) |
<a name="ab2dbdd7-5a56-4e5f-ac86-9cfc2b180346"></a>
### 8360-db9a4ab/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:32569
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp bfffb5b5ff21:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 5 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-zqbfk4cq5eep32",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 364845,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15CAE6A94711980D, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="73af5f15-021a-4cde-8b8d-e532586d6ded"></a>
### 8360-db9a4ab/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31068
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7dd5871d662f:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 9 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 54 seconds
(4/14) Running aws-sdk-ruby tests ... done in 16 seconds
(5/14) Running awscli tests ... done in 2 minutes and 21 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 0 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 25 seconds
(9/14) Running minio-go tests ... done in 2 minutes and 29 seconds
(10/14) Running minio-java tests ... done in 3 minutes and 46 seconds
(11/14) Running minio-js tests ... done in 1 minutes and 10 seconds
(12/14) Running minio-py tests ... FAILED in 5 minutes and 15 seconds
{
  "status": "FAIL",
  "name": "minio-py:test_list_objects_with_1001_files",
  "args": {
    "object_name": "84f35f20-16f3-4b40-88f2-0c7c3657bff0_0 ~ 84f35f20-16f3-4b40-88f2-0c7c3657bff0_1000",
    "bucket_name": "minio-py-test-8dc3bb79-16d2-400e-ae59-2885efa004a8"
  },
  "message": "BucketNotEmpty: message: The bucket you tried to delete is not empty.",
  "function": "list_objects(bucket_name, prefix, recursive)",
  "duration": 268279,
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1168, in test_list_objects_with_1001_files\n    client.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 416, in remove_bucket\n    self._url_open('DELETE', bucket_name=bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 1951, in _url_open\n    object_name).get_exception()\nminio.error.BucketNotEmpty: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2024, in main\n    test_list_objects_with_1001_files(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1170, in test_list_objects_with_1001_files\n    raise Exception(err)\nException: BucketNotEmpty: message: The bucket you tried to delete is not empty.\n"
}

Executed 11 out of 14 tests successfully.
```
Error: No such image: minio/minio:8360-db9a4ab

File: cmd/config-current.go [hunk 1]
```diff
@@ -234,37 +230,34 @@ func (s *serverConfig) Validate() error {
...
+		globalServerRegion = s.GetRegion()
 	}
 
-	if globalIsStorageClass {
-		s.SetStorageClass(globalStandardStorageClass, globalRRStorageClass)
+	var err error
+	if globalIsXL {
+		s.StorageClass, err = storageclass.LookupConfig(s.StorageClass, globalXLSetDriveCount)
+		if err != nil {
+			logger.FatalIf(err, "Unable to storage classes")
```
Comment (kannappanr) [inline cmd/config-current.go]:
```suggestion
			logger.FatalIf(err, "Unable to initialize storage class config")
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#7a96f75a-36f7-482e-a79d-687a79157817) |
<a name="7a96f75a-36f7-482e-a79d-687a79157817"></a>
### 8360-eb7d9b8/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30681
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c1e3fb268c72:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 50 seconds
(4/14) Running aws-sdk-ruby tests ... done in 16 seconds
(5/14) Running awscli tests ... done in 2 minutes and 10 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 17 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 30 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 28 seconds
{
  "args": {},
  "duration": 71095,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8360-eb7d9b8

File: cmd/config/storageclass/storage-class.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
+/*
+ * MinIO Cloud Storage, (C) 2017 MinIO, Inc.
```
Comment (kannappanr) [inline cmd/config/storageclass/storage-class.go]:
```suggestion
 * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
```

**Review by kannappanr — DISMISSED** [review]
LGTM except for the minor suggestion

**Review by poornas — DISMISSED** [review]
LGTM

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#aac7f98d-c800-473f-9a00-597cd3b56b89) |
<a name="aac7f98d-c800-473f-9a00-597cd3b56b89"></a>
### 8360-69c9be0/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:31048
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 11e446b47990:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 50 seconds
(4/14) Running aws-sdk-ruby tests ... done in 13 seconds
(5/14) Running awscli tests ... done in 1 minutes and 24 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 22 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 52 seconds
(9/14) Running minio-go tests ... FAILED in 4 minutes and 3 seconds
{
  "args": {},
  "duration": 103117,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Error: No such image: minio/minio:8360-69c9be0

**Review by kannappanr — APPROVED** [review]
LGTM

**Comment by harshavardhana** [comment]
PTAL @poornas

**Review by poornas — APPROVED** [review]

**Pull request merged**